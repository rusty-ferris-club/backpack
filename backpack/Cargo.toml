[package]
name = "backpack"
version = "2.1.1"
edition = "2021"


#
# == This is a dual CLI + library crate. ==
#
# * To turn this into a pure library crate, remove [[features]] and [[bin]],
#   and remove optional deps.
# * To turn into a pure CLI crate, do the same as before, but don't remove
#   optional deps, turn them into non-optional.
#
# To use both as dual crate:
#
# 1. For a CLI, nothing needs to be done.
# 2. For a library, build using `--no-default-features`, and depend using `default-features = false`
#


[dependencies]

clap = { version = "3", features = ["cargo"], optional = true }
serde = "1"
serde_json = "1"
serde_derive = "1"
serde_yaml = "^0.8.24"
serde_regex = "^1.1.0"
env_logger = "^0.9.0"
log = "^0.4.17"
anyhow = "1"
console = "^0.15.0"
url = "2"
git-url-parse = "^0.4.2"
lazy_static = "1"
regex = "1"
cached-path = "^0.5.3"
tempfile = "3"
walkdir = "2"
dirs = "4"
tracing = "^0.1.34"
merge-struct = "0.1.0"
reqwest = { version = "^0.11.11", features = ["blocking"] }
requestty = "0.4.0"
requestty-ui = "0.4.0"
tracing-forest = { version = "^0.1.4" }
tracing-subscriber = { version = "^0.3.11", features = ["env-filter"] }
# interactive-actions = "1.0.1"
# interactive-actions =  { path = "../../interactive-actions/interactive-actions" }
interactive-actions =  { git = "https://github.com/strowk/interactive-actions.git" }
tera = "^1.17.0"
tera-text-filters = "^1.0.0"
content_inspector = "0.2.4"
edit = "*"

[dev-dependencies]
insta = { version = "1.17.1", features = ["backtrace", "redactions"] }
fs_extra = "^1.2.0"
rstest = "^0.14.0"
pretty_assertions = "1"
serial_test = "0.4.0"


[features]
default = ["cli"]
# list optionals here:
cli = ["clap"]

[[bin]]
name = "bp"
required-features = ["cli"]
