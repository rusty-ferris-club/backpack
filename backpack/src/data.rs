use anyhow::Result as AnyResult;
use core::fmt::Debug;
use serde::{Deserialize, Serialize};
use url::Url;

#[derive(Clone, Debug)]
pub struct Location {
    pub url: String,
    pub domain: String,
    pub path: String,
    pub project: String,
    pub port: Option<u16>,
    pub query: Option<String>,
    pub subfolder: Option<String>,
    pub gref: Option<String>,
    pub username: String,
    pub password: Option<String>,
    pub is_git: bool,
}
impl Location {
    pub fn git_url(&self) -> String {
        format!(
            "git@{}:{}.git",
            self.domain,
            self.path.trim_start_matches('/')
        ) // needs: auth, port
    }

    pub fn web_url(&self) -> String {
        format!("https://{}{}", self.domain, self.path) // needs: auth, port
    }

    /// Create a logical `Location` from a `Url`.
    ///
    /// # Errors
    ///
    /// This function will return an error if `url` is illegal
    pub fn from(url: &Url, is_git: bool) -> AnyResult<Self> {
        let path = url.path();
        let parts = path.split("/-/").collect::<Vec<_>>();
        let (path, subfolder) = if parts.len() == 2 {
            (parts[0], Some(parts[1].to_string()))
        } else {
            (path, None)
        };

        let project = path
            .split('/')
            .last()
            .ok_or_else(|| anyhow::anyhow!("cannot find project name"))?;

        Ok(Self {
            url: url.to_string(),
            domain: url.domain().unwrap_or_default().to_string(),
            path: path.trim_end_matches(".git").to_string(),
            project: project.to_string(),
            port: url.port(),
            query: url.query().map(str::to_string),
            subfolder,
            gref: url.fragment().map(str::to_string),
            username: url.username().to_string(),
            password: url.password().map(str::to_string),
            is_git,
        })
    }
}

#[derive(Debug)]
pub struct Archive {
    pub url: String,
    pub root: ArchiveRoot,
}

#[derive(Debug)]
pub enum ArchiveRoot {
    Folder(String),
    FirstFolder,
    None,
}
#[derive(Debug)]
pub struct Assets {
    pub archive: Option<Archive>,
    pub git: Option<String>,
}

pub enum Overwrite {
    Ask,
    Always,
    Never,
    Custom(Box<dyn Fn(&str) -> bool>),
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
pub enum CopyMode {
    #[serde(rename = "new")]
    Copy,
    #[serde(rename = "apply")]
    Apply,
    #[serde(rename = "all")]
    All,
}

#[allow(clippy::struct_excessive_bools)]
#[derive(Debug)]
pub struct Opts {
    pub show_progress: bool,
    pub overwrite: bool,
    pub mode: CopyMode,
    pub is_git: bool,
    pub no_cache: bool,
    pub always_yes: bool,
    pub remote: Option<String>,
}

#[cfg(test)]
mod tests {
    use super::*;
    use pretty_assertions::assert_eq;

    #[test]
    fn test_location_urls() {
        let loc = Location::from(
            &Url::parse("https://github.com/user/repo.git").unwrap(),
            true,
        )
        .unwrap();
        assert_eq!(loc.git_url(), "git@github.com:user/repo.git");
        assert_eq!(loc.web_url(), "https://github.com/user/repo");

        let loc =
            Location::from(&Url::parse("https://github.com/user/repo").unwrap(), false).unwrap();
        assert_eq!(loc.git_url(), "git@github.com:user/repo.git");
        assert_eq!(loc.web_url(), "https://github.com/user/repo");

        let loc = Location::from(
            &Url::parse("https://github.com/user/repo.git").unwrap(),
            false,
        )
        .unwrap();
        assert_eq!(loc.git_url(), "git@github.com:user/repo.git");
        assert_eq!(loc.web_url(), "https://github.com/user/repo");
    }
}
