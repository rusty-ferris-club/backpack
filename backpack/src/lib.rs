//#![deny(clippy::pedantic)]
//#![deny(clippy::nursery)]
#![allow(clippy::must_use_candidate)]
#![allow(clippy::module_name_repetitions)]

pub mod config;
pub mod content;
pub mod data;
pub mod fetch;
pub mod git;
mod merge;
pub mod run;
pub mod shortlink;
pub mod vendors;
