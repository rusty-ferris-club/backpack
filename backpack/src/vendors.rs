use crate::{
    config::{CustomVendor, VendorsConfig},
    data::{Archive, ArchiveRoot, Assets, Location},
    git::GitProvider,
};
use anyhow::Result as AnyResult;
use core::fmt::Debug;
use tracing;
pub struct Vendors<'a> {
    config: Option<&'a VendorsConfig>,
}

impl<'a> Vendors<'a> {
    pub const fn new(config: Option<&'a VendorsConfig>) -> Self {
        Self { config }
    }

    #[tracing::instrument(name = "lookup_vendor", skip(self), err)]
    pub fn lookup(&self, vendor: &str) -> AnyResult<Box<dyn Vendor>> {
        let github = CustomVendor {
            kind: "github".to_string(),
            base: "github.com".to_string(),
        };
        let v = if vendor.is_empty() {
            self.config
                .and_then(|v| v.vendors_default.as_ref())
                .or(Some(&github))
        } else {
            self.config
                .and_then(|vendors| vendors.custom.as_ref())
                .and_then(|h| h.get(vendor))
        };
        v.map_or_else(
            || Vendors::lookup_table(vendor, None),
            |v| Vendors::lookup_table(&v.kind, Some(v.base.as_ref())),
        )
    }
    fn lookup_table(token: &str, base: Option<&str>) -> AnyResult<Box<dyn Vendor>> {
        match token {
            "gh" | "github.com" | "github" => Ok(Box::new(Github::new(base))),
            "gist.github.com" | "gist" => Ok(Box::new(GithubGist::new(base))),
            "gl" | "gitlab.com" | "gitlab" => Ok(Box::new(Gitlab::new(base))),
            "bb" | "bitbucket.org" | "bitbucket" => Ok(Box::new(BitBucket::new(base))),
            _ => anyhow::bail!("no vendor found for: {}", token),
        }
    }
}

pub trait Vendor {
    fn base(&self) -> &str;

    /// Resolve a location's assets.
    /// For example, a location my describe a github repo, and a branch, and its assets
    /// will be where you can download the code of that branch as a tarball.
    ///
    /// # Errors
    ///
    /// This function will return an error if network or other I/O is erroring.
    fn resolve(&self, location: &Location, git: &dyn GitProvider) -> AnyResult<Assets>;
}

impl Debug for dyn Vendor {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        write!(f, "{}", self.base())
    }
}

// add gist vendor
// the new release should include single file support + gist
// find a way to test single file cloning

pub struct Github {
    base: String,
}

impl Github {
    pub fn new(base: Option<&str>) -> Self {
        Self {
            base: base.map_or_else(|| "github.com".to_string(), ToString::to_string),
        }
    }
}
impl Vendor for Github {
    fn base(&self) -> &str {
        self.base.as_str()
    }
    #[tracing::instrument(name = "github_resolve", skip_all, err)]
    fn resolve(&self, location: &Location, git: &dyn GitProvider) -> AnyResult<Assets> {
        let gref = git.get_ref_or_default(location)?.ref_;
        Ok(Assets {
            archive: Some(Archive {
                url: format!(
                    "https://{}{}/archive/{}.tar.gz",
                    self.base(),
                    location.path,
                    gref,
                ),
                root: ArchiveRoot::FirstFolder,
            }),
            git: Some(format!(
                "git@{}:{}.git",
                self.base(),
                location.path.trim_start_matches('/')
            )),
        })
    }
}

pub struct LocalGit {}

impl Vendor for LocalGit {
    fn base(&self) -> &str {
        "file"
    }
    #[tracing::instrument(name = "localgit_resolve", skip_all, err)]
    fn resolve(&self, location: &Location, _git: &dyn GitProvider) -> AnyResult<Assets> {
        Ok(Assets {
            archive: None,
            git: Some(location.url.clone()),
        })
    }
}

pub struct GithubGist {
    base: String,
}

impl GithubGist {
    pub fn new(base: Option<&str>) -> Self {
        Self {
            base: base.map_or_else(|| "gist.github.com".to_string(), ToString::to_string),
        }
    }
}
impl Vendor for GithubGist {
    fn base(&self) -> &str {
        self.base.as_str()
    }
    #[tracing::instrument(name = "github_gist_resolve", skip_all, err)]
    fn resolve(&self, location: &Location, git: &dyn GitProvider) -> AnyResult<Assets> {
        let refs = git.ls_remote(location)?;
        let head = refs
            .iter()
            .find(|r| r.ref_ == "HEAD")
            .ok_or_else(|| anyhow::anyhow!("no HEAD ref found"))?;

        Ok(Assets {
            archive: Some(Archive {
                url: format!(
                    //   https://gist.github.com/jondot/15086f59dab44f30bb10f82ca09f4887/archive/44a751f50ea93f92c2bc6332e4de770429862888.zip
                    "https://{}{}/archive/{}.tar.gz",
                    self.base(),
                    location.path,
                    head.ref_,
                ),
                root: ArchiveRoot::FirstFolder,
            }),
            git: Some(format!(
                "git@{}:{}.git",
                self.base(),
                location.path.trim_start_matches('/')
            )),
        })
    }
}
pub struct BitBucket {
    base: String,
}

impl BitBucket {
    pub fn new(base: Option<&str>) -> Self {
        Self {
            base: base.map_or_else(|| "bitbucket.org".to_string(), ToString::to_string),
        }
    }
}

impl Vendor for BitBucket {
    fn base(&self) -> &str {
        self.base.as_str()
    }

    #[tracing::instrument(name = "bitbucket_resolve", skip_all, err)]
    fn resolve(&self, location: &Location, git: &dyn GitProvider) -> AnyResult<Assets> {
        let gref = git.get_ref_or_default(location)?.ref_;
        Ok(Assets {
            archive: Some(Archive {
                url: format!(
                    "https://{}{}/get/{}.tar.gz",
                    self.base(),
                    location.path,
                    gref,
                ),
                root: ArchiveRoot::FirstFolder,
            }),
            git: Some(format!(
                "git@{}:{}.git",
                self.base(),
                location.path.trim_start_matches('/')
            )),
        })
    }
}

pub struct Gitlab {
    base: String,
}

impl Gitlab {
    pub fn new(base: Option<&str>) -> Self {
        Self {
            base: base.map_or_else(|| "gitlab.com".to_string(), ToString::to_string),
        }
    }
}

impl Vendor for Gitlab {
    fn base(&self) -> &str {
        self.base.as_str()
    }

    #[tracing::instrument(name = "gitlab_resolve", skip_all, err)]
    fn resolve(&self, location: &Location, git: &dyn GitProvider) -> AnyResult<Assets> {
        let gref = git.get_ref_or_default(location)?.ref_;
        let gref_file = gref.replace('/', "-");
        Ok(Assets {
            archive: Some(Archive {
                url: format!(
                    "https://{}{}/-/archive/{}/{}-{}.tar.gz",
                    self.base(),
                    location.path,
                    gref,
                    location.project,
                    gref_file,
                ),
                root: ArchiveRoot::FirstFolder,
            }),
            git: Some(format!(
                "git@{}:{}.git",
                self.base(),
                location.path.trim_start_matches('/')
            )),
        })
    }
}
