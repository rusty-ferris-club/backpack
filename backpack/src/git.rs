use crate::data::Location;
use anyhow::{bail, Context, Result};
use std::process::Command;
use tracing;

pub trait GitProvider {
    /// Performs a shallow Git clone.
    ///
    /// # Errors
    ///
    /// This function will return an error if underlying Git provider failed.
    fn shallow_clone(&self, location: &Location, out: &str) -> Result<()>;

    /// Perform a Git ls-remote on a remote location.
    ///
    /// # Errors
    ///
    /// This function will return an error if underlying ls-remote implementation failed.
    fn ls_remote(&self, location: &Location) -> Result<Vec<RemoteInfo>>;

    /// Get a Git ref from a remote, for a given logical branch in `locaation`.
    ///
    /// # Errors
    ///
    /// This function will return an error if underlying remote resolving implementation failed.
    fn get_ref_or_default(&self, location: &Location) -> Result<RemoteInfo>;

    /// Get the current repo's main remote url
    ///
    /// # Errors
    ///
    /// This function will return an error if underlying remote resolving implementation failed.
    fn get_local_url(&self) -> Result<String>;
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct RemoteInfo {
    pub revision: String,
    pub ref_: String,
}

#[derive(Default)]
pub struct GitCmd {}
impl GitProvider for GitCmd {
    #[tracing::instrument(name = "git_get_ref", skip_all, err)]
    fn get_ref_or_default(&self, location: &Location) -> Result<RemoteInfo> {
        let refs = self.ls_remote(location)?;
        if let Some(ref gref) = location.gref {
            let ref_ = refs
                .iter()
                .find(|r| r.ref_.ends_with(gref))
                .ok_or_else(|| anyhow::anyhow!("no such ref found: {}", gref))?;
            return Ok(ref_.clone());
        }

        let head = refs
            .iter()
            .find(|r| r.ref_ == "HEAD")
            .ok_or_else(|| anyhow::anyhow!("no HEAD ref found"))?;
        let default_branch = refs
            .iter()
            .find(|r| r.ref_ != "HEAD" && r.revision == head.revision)
            .ok_or_else(|| anyhow::anyhow!("no default branch found"))?;
        Ok(default_branch.clone())
    }

    #[tracing::instrument(name = "git_ls_remote", skip_all, err)]
    fn ls_remote(&self, location: &Location) -> Result<Vec<RemoteInfo>> {
        let remote = if location.is_git {
            location.git_url()
        } else {
            location.web_url()
        };
        let rremote = remote.as_str();

        let process = Command::new("git")
            .arg("ls-remote")
            .arg(rremote)
            .output()
            .with_context(|| format!("cannot run git ls-remote on '{rremote}'"))?;
        if !process.status.success() {
            log::error!("git ls-remote failed. stderr output:");
            let err = String::from_utf8_lossy(&process.stderr);
            err.split('\n').for_each(|line| log::error!("> {}", line));
            anyhow::bail!(
                "git ls-remote '{}' failed with exit code {}\n---\n{}",
                rremote,
                process
                    .status
                    .code()
                    .map_or_else(|| "None".into(), |code| code.to_string()),
                err
            );
        }

        String::from_utf8_lossy(&process.stdout)
            .split('\n')
            .filter(|line| !line.is_empty())
            .map(|line| {
                let (revision, ref_) = line
                    .split_once('\t')
                    .ok_or_else(|| anyhow::format_err!("Output line contains no '\\t'"))?;
                anyhow::ensure!(
                    !ref_.contains('\t'),
                    "Output line contains more than one '\\t'"
                );
                Ok(RemoteInfo {
                    revision: revision.into(),
                    ref_: ref_.into(),
                })
            })
            .collect::<Result<Vec<RemoteInfo>>>()
    }

    #[tracing::instrument(name = "git_clone", skip_all, err)]
    fn shallow_clone(&self, location: &Location, out: &str) -> anyhow::Result<()> {
        // -b branch
        // need to take location, and ask it for the git url
        let giturl = location.git_url();
        let branch = location.gref.as_deref();
        let mut args = vec!["clone", "--depth=1"];
        if let Some(branch) = branch {
            args.push("-b");
            args.push(branch);
        }
        args.push(&giturl);
        args.push(out);

        let output = Command::new("git")
            .args(&args)
            .output()
            .context("git handling failed")?;

        if !output.status.success() {
            anyhow::bail!(
                "cannot clone: {}\n---\n{}",
                giturl,
                String::from_utf8_lossy(&output.stderr[..])
            );
        }
        Ok(())
    }

    fn get_local_url(&self) -> anyhow::Result<String> {
        let process = Command::new("git")
            .arg("remote")
            .arg("get-url")
            .arg("origin")
            .output()
            .with_context(|| "cannot run git remote on local repo".to_string())?;

        if !process.status.success() {
            anyhow::bail!(
                "cannot find local url: {}",
                String::from_utf8_lossy(&process.stderr[..])
            );
        }

        let url = String::from_utf8_lossy(&process.stdout[..]);
        let lines = url.lines().collect::<Vec<_>>();
        if lines.len() != 1 {
            // too many urls, cannot decide
            bail!("repo has more than one remote URL")
        }
        match lines.first() {
            Some(ln) => Ok((*ln).to_string()),
            _ => bail!("no repo remote URL found"),
        }
    }
}
