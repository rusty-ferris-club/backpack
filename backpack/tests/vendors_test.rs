use backpack::data::Location;
use backpack::git::{GitProvider, RemoteInfo};
use backpack::vendors::{BitBucket, Github, Gitlab, Vendor};
use insta::assert_debug_snapshot;
use pretty_assertions::assert_eq;
use reqwest::{self, StatusCode};
use url::Url;

struct TestGitProvider {
    remote: RemoteInfo,
}
impl GitProvider for TestGitProvider {
    fn get_ref_or_default(
        &self,
        _location: &Location,
    ) -> anyhow::Result<backpack::git::RemoteInfo> {
        Ok(self.remote.clone())
    }
    fn ls_remote(&self, _location: &Location) -> anyhow::Result<Vec<backpack::git::RemoteInfo>> {
        Ok(vec![])
    }
    fn shallow_clone(&self, _location: &Location, _out: &str) -> anyhow::Result<()> {
        Ok(())
    }
    fn get_local_url(&self) -> anyhow::Result<String> {
        Ok(String::new())
    }
}

#[test]
fn test_github() {
    let git = TestGitProvider {
        remote: RemoteInfo {
            revision: "rev".to_string(),
            ref_: "refs/heads/master".to_string(),
        },
    };
    let gh = Github::new(None);
    let location = Location::from(
        &Url::parse("https://github.com/jondot/hygen").unwrap(),
        false,
    )
    .unwrap();
    let assets = gh.resolve(&location, &git).unwrap();
    assert_debug_snapshot!(assets);
    assert_eq!(
        reqwest::blocking::get(assets.archive.unwrap().url)
            .unwrap()
            .status(),
        StatusCode::OK
    );
}

#[test]
fn test_gitlab() {
    let git = TestGitProvider {
        remote: RemoteInfo {
            revision: "rev".to_string(),
            ref_: "refs/heads/master".to_string(),
        },
    };
    let gh = Gitlab::new(None);
    let location = Location::from(
        &Url::parse("https://gitlab.com/jondot/backpack-e2e").unwrap(),
        false,
    )
    .unwrap();
    let assets = gh.resolve(&location, &git).unwrap();
    assert_debug_snapshot!(assets);
    assert_eq!(
        reqwest::blocking::get(assets.archive.unwrap().url)
            .unwrap()
            .status(),
        StatusCode::OK
    );
}

#[test]
fn test_bitbucket() {
    let git = TestGitProvider {
        remote: RemoteInfo {
            revision: "rev".to_string(),
            ref_: "refs/heads/master".to_string(),
        },
    };
    let gh = BitBucket::new(None);
    let location = Location::from(
        &Url::parse("https://bitbucket.org/microsoft/azure-cli-run").unwrap(),
        false,
    )
    .unwrap();
    let assets = gh.resolve(&location, &git).unwrap();
    assert_debug_snapshot!(assets);
    assert_eq!(
        reqwest::blocking::get(assets.archive.unwrap().url)
            .unwrap()
            .status(),
        StatusCode::OK
    );
}
