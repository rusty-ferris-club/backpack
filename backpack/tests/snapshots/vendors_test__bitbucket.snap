---
source: backpack/tests/vendors_test.rs
expression: assets
---
Assets {
    archive: Some(
        Archive {
            url: "https://bitbucket.org/microsoft/azure-cli-run/get/refs/heads/master.tar.gz",
            root: FirstFolder,
        },
    ),
    git: Some(
        "git@bitbucket.org:microsoft/azure-cli-run.git",
    ),
}
